hpx/parallel/util/loop.hpp

```diff
@@ -68,12 +69,176 @@ namespace hpx { namespace parallel { namespace util
+                base = base + (rhs*chunk_size);
+                return *this;
+            }
+
+            inline prefetching_iterator& operator-=(difference_type rhs)
+            {
+                idx = idx - (rhs*chunk_size);
+                base = base - (rhs*chunk_size);
+                return *this;
+            }
+
+            inline prefetching_iterator& operator++()
+            {
+                idx = idx + chunk_size;
+                base = base + chunk_size;
+                return *this;
+            }
+
+            inline prefetching_iterator& operator--()
+            {
+                idx = idx - chunk_size;
+                base = base - chunk_size;
+                return *this;
+            }
+
+            inline prefetching_iterator operator++(int)
+            {
+                prefetching_iterator tmp(*this);
+                operator++();
+                return tmp;
+            }
+
+            inline prefetching_iterator operator--(int)
+            {
+                prefetching_iterator tmp(*this);
+                operator--();
+                return tmp;
+            }
+
+            inline difference_type operator-(const prefetching_iterator& rhs) const
+            {
+                return (idx-rhs.idx)/chunk_size;
+            }
+
+            inline prefetching_iterator operator+(difference_type rhs) const
+            {
+                return prefetching_iterator((idx+(rhs*chunk_size)),(base+(rhs*chunk_size)),chunk_size,range_size,M_);
+            }
+
+            inline prefetching_iterator operator-(difference_type rhs) const
+            {
+                return prefetching_iterator((idx-(rhs*chunk_size)),(base-(rhs*chunk_size)),chunk_size,range_size,M_);
+            }
+
+            friend inline prefetching_iterator operator+(difference_type lhs, const prefetching_iterator& rhs)
+            {
+                return rhs + lhs;
+            }
+
+            friend inline prefetching_iterator operator-(difference_type lhs, const prefetching_iterator& rhs)
+            {
+                return lhs - rhs;
+            }
+
+
+            inline bool operator==(const prefetching_iterator& rhs) const {return idx == rhs.idx;}
+            inline bool operator!=(const prefetching_iterator& rhs) const {return idx != rhs.idx;}
+            inline bool operator>(const prefetching_iterator& rhs) const {return idx > rhs.idx;}
+            inline bool operator<(const prefetching_iterator& rhs) const {return idx < rhs.idx;}
+            inline bool operator>=(const prefetching_iterator& rhs) const {return idx >= rhs.idx;}
+            inline bool operator<=(const prefetching_iterator& rhs) const {return idx <= rhs.idx;}
+
+
+            inline std::size_t operator*() const {return idx;}
+        };
+
+
+        //Helper class to initialize prefetching_iterator
+        template<typename T>
+        struct prefetcher_context
+        {
+            std::vector<std::size_t> range;
```

Comment:
The range of indexes shouldn't be stored as a vector. Could you use boost::irange instead?